src/selvage/test/flow.clj

```diff
@@ -0,0 +1,199 @@
+  [& body]
+  `(update-report-counters
+     (binding [*intermediate-report* (or (some-> t/*report-counters* deref) ~t/*initial-report-counters*)]
+       (binding [t/*report-counters* (ref *intermediate-report*)]
+         ~@body
+         (diff-test-result *intermediate-report* @t/*report-counters*)))))
+
+(defn- test-counter-reset
+  "Grab current test stats and return clojure that resets to those states and
+   runs function. Used for retrying function execution."
+  [f]
+  (let [output-counters-before @t/*report-counters*]
+    (fn [& args]
+      (dosync (ref-set t/*report-counters* output-counters-before))
+      (apply f args))))
+
+(defn retry [f]
+  (letfn [(retry? [elapsed-millis] (<= elapsed-millis *probe-timeout*))
+          (retry-f [elapsed-so-far f w]
+            (let [[time [success? desc :as res]] (core/timed-apply f w)
+                  elapsed                        (+ elapsed-so-far time)]
+              (if success?
+                res
+                (if (retry? elapsed)
+                  (do
+                    (Thread/sleep *probe-sleep-period*)
+                    ;; time accounting might be improved
+                    (retry-f (+ elapsed *probe-sleep-period*) f w))
+                  [false desc]))))]
+    (partial retry-f 0
+      (test-counter-reset f))))
+
+(defn run-test-var [test-var]
+  `(fn [] (t/test-var (var ~test-var))))
+
+(defn run-test-expr [testing-expr]
+  `(fn [] (do ~testing-expr)))
+
+(defn report-diff->successful? [report-before report-after]
+  (and (zero? (- (:fail report-before 0)
+                 (:fail report-after 0)))
+       (zero? (- (:error report-before 0)
+                 (:error report-after 0)))))
+
+(defn test->fn-expr [run-test-fn]
+  `(fn [world#]
+     (let [writer#         (new StringWriter)
+           report-before# @t/*report-counters*]
+       (binding [*world*                 world#
+                 clojure.test/*test-out* writer#]
+         (~run-test-fn)
+         (let [result#  (report-diff->successful? report-before# @t/*report-counters*)
+               success# (when result#
+                          world#)]
+           [success# (str writer#)])))))
+
+(defn- is-test-var? [form]
+  (and (symbol? form)
+       (-> form resolve meta :test)))
+(defn- is-testing? [form]
+  (and (coll? form)
+       (-> form first name (= "testing"))))
+(defn- is-check-ref? [form]
+  (and (coll? form)
+       (-> form first name (= "testing"))))
+
+(defn- classify [form]
+  (cond
+    (is-testing? form)         [:check
+                                (-> form run-test-expr test->fn-expr)
+                                (str form)]
+    (is-test-var? form)        [:check
+                                (-> form run-test-var test->fn-expr)
+                                (str form)]
+    (core/is-query? form)      [:query
+                                (core/transition->fn-expr form)
+                                (str form)]
+    (core/is-transition? form) [:transition
```

Comment:
this check is new and causes more helpful failures when people doing things like `(defflow the-flow 1)`